include/envoy/http/header_map.h

```diff
@@ -637,12 +631,90 @@ class HeaderMap {
+ * compilation units and will still point to the same slot. Headers are registered independently
+ * for each concrete header map type and do not overlap. Handles are strongly typed and do not
+ * allow mixing.
+ */
+template <class T> class CustomInlineHeaderRegistry {
+public:
+  using RegistrationMap = std::map<LowerCaseString, size_t>;
+  // Phantom is used here to force the compiler to verify that handles are not mixed between
+  // concrete header map types.
+  using Handle = Phantom<RegistrationMap::const_iterator, T>;
+
+  /**
+   * Register an inline header and return a handle for use in inline header calls. Must be called
+   * prior to finalize().
+   */
+  static Handle registerInlineHeader(const LowerCaseString& header_name) {
+    static size_t inline_header_index = 0;
+
+    ASSERT(!mutableFinalized());
+    auto entry = headerNameToIndex().find(header_name);
+    if (entry == headerNameToIndex().end()) {
+      headerNameToIndex()[header_name] = inline_header_index++;
+    }
+    return Handle(headerNameToIndex().find(header_name));
+  }
+
+  /**
+   * Fetch the handle for a registered inline header. May only be called after finalized().
+   */
+  static absl::optional<Handle> getInlineHeader(const LowerCaseString& header_name) {
+    ASSERT(mutableFinalized());
+    auto entry = headerNameToIndex().find(header_name);
+    if (entry != headerNameToIndex().end()) {
+      return Handle(entry);
+    }
+    return absl::nullopt;
+  }
+
+  /**
+   * Fetch all registered headers. May only be called after finalized().
+   */
+  static const RegistrationMap& headers() {
+    ASSERT(mutableFinalized());
+    return headerNameToIndex();
+  }
+
+  /**
+   * Finalize the custom header registrations. No further changes are allowed after this point.
+   * This guaranteed that all header maps created by the process have the same variable size and
+   * custom registrations.
+   */
+  static void finalize() {
+    ASSERT(!mutableFinalized());
+    mutableFinalized() = true;
+  }
+
+private:
+  static RegistrationMap& headerNameToIndex() { MUTABLE_CONSTRUCT_ON_FIRST_USE(RegistrationMap); }
+  static bool& mutableFinalized() { MUTABLE_CONSTRUCT_ON_FIRST_USE(bool); }
+};
+
+/**
+ * Static initializer to register a custom header in a compilation unit. This can be used by
+ * extensions to register custom headers.
+ */
+template <class T> class RegisterCustomInlineHeader {
+public:
+  RegisterCustomInlineHeader(const LowerCaseString& header)
+      : handle_(CustomInlineHeaderRegistry<T>::registerInlineHeader(header)) {}
+
+  typename CustomInlineHeaderRegistry<T>::Handle handle() { return handle_; }
+
+private:
+  const typename CustomInlineHeaderRegistry<T>::Handle handle_;
+};
+
 /**
  * Typed derived classes for all header map types.
  */
 
 // Base class for both request and response headers.
-class RequestOrResponseHeaderMap : public virtual HeaderMap {
+class RequestOrResponseHeaderMap : public HeaderMap {
```

Comment:
the de-virtualization of this inheritance is great; but I wonder if you could break this into a separate PR that could be merged without controversy or extensive benchmarking, thus simplifying this one. WDYT?

Reply:
Yes I can likely split this into a different PR, but I would rather decide on whether we want to go in this direction or not first, since I'm not sure if it's worth it by itself (given that it introduces it's own templating complexity.)